File: backend/app/app/models/student.py
from typing import TYPE_CHECKING, Any

from sqlalchemy import Boolean, Column, Integer, String, Float, Text
from sqlalchemy.dialects.postgresql.base import UUID
from sqlalchemy.orm import relationship
from sqlalchemy.sql.schema import ForeignKey
from sqlalchemy.sql.sqltypes import ARRAY
import uuid
from app.db.base_class import Base


class Student(Base):
    uuid = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    num_carte = Column(String, unique=True, primary_key=True)
    num_select = Column(String, nullable=True, unique=True)
    last_name = Column(String)
    first_name = Column(String)
    date_birth = Column(String)
    place_birth = Column(String)
    address = Column(String)
    sex = Column(String)
    situation = Column(String)
    telephone = Column(String)
    nation = Column(String)
    num_cin = Column(String)
    date_cin = Column(String)
    place_cin = Column(String)
    type = Column(String) # passant , redoublant
    receipt_list = Column(ARRAY(String))
    receipt = Column(String)
    photo = Column(String)
    baccalaureate_num = Column(String)
    baccalaureate_center = Column(String)
    baccalaureate_years = Column(String)
    baccalaureate_series = Column(String)
    work = Column(String)

    father_name = Column(String)
    father_work = Column(String)
    mother_name = Column(String)
    mother_work = Column(String)
    parent_address = Column(String)

    level = Column(String)  # niveau
    inf_semester = Column(String)
    sup_semester = Column(String)

    actual_years = Column(ARRAY(String))
    enter_years = Column(String)

    uuid_journey = Column(UUID(as_uuid=True), ForeignKey("journey.uuid"))
    uuid_mention = Column(UUID(as_uuid=True), ForeignKey("mention.uuid"))

    is_selected = Column(Boolean, default=False)
    mean = Column(Float)

    journey = relationship("Journey", foreign_keys=[uuid_journey])
    mention = relationship("Mention", foreign_keys=[uuid_mention])

File: frontend/src/app/models/student.ts
import { NzTableSortOrder, NzTableSortFn, NzTableFilterList, NzTableFilterFn } from "ng-zorro-antd/table"
import { Journey } from "./journey"
import { Receipt } from "./receipt"

export interface AncienStudent{
    uuid: string,
    last_name: string,
    first_name: string,
    date_birth: Date,
    place_birth: string,
    address: string,
    sex: string,
    nation: string,
    num_cin: string,
    date_cin: Date,
    place_cin: string,
    mention: string,
    actual_years: string,
    num_carte: string,
    num_baccalaureat: string,
    baccalaureate_years: string,
    mean: number,
    inf_semester: string,
    sup_semester: string,
    price_rigth: number,
    type: string,
    photo: string,
    num_receipt: Array<string>
    date_receipt: Array<string>
    journey: Journey,
    receipt: Receipt

    num_select: string,
    is_selected: false,
    enter_years: string,

    situation: string,
    telephone: string,
    baccalaureate_num: string,
    baccalaureate_center: string,
    baccalaureate_series: string,
    work: string,
    father_name: string,
    father_work: string,
    mother_name: string,
    mother_work: string,
    parent_address: string,
    uuid_mention: string,
    level: string,
}

export interface StudentColumn{
    title: string;
    last_name: string,
    inf_semester: string,
    sup_semester: string,
    journey: Journey,
    is_selected: false,
    level: string,
}

export interface StudentInfo{
    info: any;
    Normal: any,
    Rattrapage: any,
}

export interface ColumnItem{
    name: string;
    sortOrder: NzTableSortOrder | null;
    sortFn: NzTableSortFn<StudentColumn> | null;
    listOfFilter: NzTableFilterList;
    filterFn: NzTableFilterFn<StudentColumn> | null;
    filterMultiple: boolean;
    sortDirections: NzTableSortOrder[];
}